stage1_fly/run/main.go

```diff
@@ -199,19 +200,40 @@ func evaluateMounts(rfs string, app string, p *stage1commontypes.Pod) ([]flyMoun
 			)
 		}
 
-		var flags uintptr = syscall.MS_BIND
+		var (
+			flags     uintptr = syscall.MS_BIND
+			recursive         = tuple.V.Recursive != nil && *tuple.V.Recursive
+			ro                = tuple.V.ReadOnly != nil && *tuple.V.ReadOnly
+		)
+
 		// If Recursive is not set, default to non-recursive.
-		if tuple.V.Recursive != nil && *tuple.V.Recursive {
+		if recursive {
 			flags |= syscall.MS_REC
 		}
+
 		argFlyMounts = append(argFlyMounts,
 			flyMount{tuple.V.Source, rfs, tuple.M.Path, "none", flags},
 		)
 
-		if tuple.V.ReadOnly != nil && *tuple.V.ReadOnly {
+		if ro {
 			argFlyMounts = append(argFlyMounts,
 				flyMount{"", rfs, tuple.M.Path, "none", flags | syscall.MS_REMOUNT | syscall.MS_RDONLY},
 			)
+
+			if recursive {
+				// Every sub-mount needs to be remounted read-only separately
+				mnts, err := stage0.GetMountsForPrefix(tuple.V.Source + "/")
+				if err != nil {
+					return nil, errwrap.Wrap(fmt.Errorf("error getting mounts under %q from mountinfo", tuple.V.Source), err)
+				}
+
+				for _, mnt := range mnts {
+					innerRelPath := tuple.M.Path + strings.Replace(mnt.MountPoint, tuple.V.Source, "", -1)
+					argFlyMounts = append(argFlyMounts,
+						flyMount{"", rfs, innerRelPath, "none", flags | syscall.MS_REMOUNT | syscall.MS_RDONLY},
```

Comment:
man mount(2) says about MS_REMOUNT:

> The mountflags and data arguments should match the values used in the original mount() call, except for  those  parameters that are being deliberately changed.

So MS_BIND should be passed.

Also, [do_remount](https://github.com/torvalds/linux/blob/v4.7/fs/namespace.c#L2215) (performing the MS_REMOUNT) behaves differently if the flag MS_BIND is not given.

In my manual test above, `/tmp/data/RW` was a bind mount. But in the functional test, `tmpdir2path` (/tmp/rkt-tests-flyxxxx/dir2) is a new mount (tmpfs). Maybe that's the difference explaining why the tests passed? I don't know.